### PR TITLE
[WIP] Adding support for `flax` for `pipelines`.

### PR DESCRIPTION
**Pull request by Narsil**
# What does this PR do?

Fixes # (issue)

```python
from transformers import FlaxRobertaModel, RobertaTokenizerFast,
RobertaForMaskedLM, pipeline
import numpy as np
import datetime
import tqdm
import random

def data(n):
    for _ in range(n):
        yield "JAX/Flax is amazing <mask>"

def flax(n):
    print("----")
    print("Flax")
    start = datetime.datetime.now()
    pipe = pipeline(
        model="distilbert-base-uncased-finetuned-sst-2-english",
        device=0,
        framework="flax",
        model_kwargs={"from_pt": True},
    )
    print("Loading flax", datetime.datetime.now() - start)

    for out in tqdm.tqdm(pipe(data(n), batch_size=512)):
        pass

def tf(n):
    print("----")
    print("TF")
    start = datetime.datetime.now()
    pipe =
pipeline(model="distilbert-base-uncased-finetuned-sst-2-english",
device=0, framework="tf")
    print("Loading TF", datetime.datetime.now() - start)

    for out in tqdm.tqdm(pipe(data(n), batch_size=512)):
        pass

def pt(n):
    print("----")
    print("PT")
    start = datetime.datetime.now()
    pipe =
pipeline(model="distilbert-base-uncased-finetuned-sst-2-english",
device=0, framework="pt")
    print("Loading PT", datetime.datetime.now() - start)

    for out in tqdm.tqdm(pipe(data(n), batch_size=512)):
        pass
        # print(out)

if __name__ == "__main__":
    n = 20000
    # pt(n)
    # flax(n)
    tf(n)
```

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/master/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [ ] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/master/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/master/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by patrickvonplaten** [comment]
@Narsil - please ping me once you would like to have a review on this :-)

**Comment by Narsil** [comment]
@patrickvonplaten I think this should be ready for a first pass review.

Let's focus on the big picture and overall architecture for this. The main pain points I am thinking of:

- Should we force padding to get compilation cache hit on jax.
- Should we enable JIT automatically or not? (If we want `pjit` or others maybe we should delegate that choice to users ?)

File: src/transformers/feature_extraction_utils.py [hunk 1]
```diff
@@ -121,6 +121,9 @@ def convert_to_tensors(self, tensor_type: Optional[Union[str, TensorType]] = Non
 
         # Convert to TensorType
         if not isinstance(tensor_type, TensorType):
+            if tensor_type == "flax":
```
Comment (patrickvonplaten) [inline src/transformers/feature_extraction_utils.py:124]: Do we need that? I'm a bit worried that this aliasing doesn't exist in other spots and that it'll trigger some bugs from users

File: src/transformers/generation_flax_utils.py [hunk 2]
```diff
@@ -237,12 +241,23 @@ def generate(
...
+                    seq_len = 1
+                else:
+                    seq_len = input_ids.shape[1]
+                max_length = seq_len + max_new_tokens
         bos_token_id = bos_token_id if bos_token_id is not None else self.config.bos_token_id
         pad_token_id = pad_token_id if pad_token_id is not None else self.config.pad_token_id
         eos_token_id = eos_token_id if eos_token_id is not None else self.config.eos_token_id
         decoder_start_token_id = (
-            decoder_start_token_id if decoder_start_token_id else self.config.decoder_start_token_id
+            decoder_start_token_id if decoder_start_token_id is not None else self.config.decoder_start_token_id
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:260]: thanks!

File: src/transformers/generation_flax_utils.py [hunk 3]
```diff
@@ -237,12 +241,23 @@ def generate(
             >>> print("Generated:", tokenizer.batch_decode(outputs, skip_special_tokens=True))
         """
         # set init values
-        max_length = max_length if max_length is not None else self.config.max_length
+        if max_length is not None and max_new_tokens is not None:
+            raise ValueError("Both `max_new_tokens` and `max_length` are defined which is incorrect, use only one.")
+        if max_length is None:
```
Comment (patrickvonplaten) [inline src/transformers/generation_flax_utils.py:246]: Could we reduce this to two lines?

```suggestion
        prompt_length = 1 if self.config.is_encoder_decoder else input_ids.shape[1]
        max_length = prompt_length + max_new_tokens if max_new_tokens is not None else self.config.max_length
```
```

Comment (patrickvonplaten) [reply src/transformers/generation_flax_utils.py:246]: IMO this is a bit more readable and would avoid a double nested if - else

File: src/transformers/models/big_bird/modeling_flax_big_bird.py [hunk 4]
```diff
@@ -1445,6 +1445,42 @@ def setup(self):
             dtype=self.dtype,
         )
 
+    def _pad_to_block_size(
```
Comment (patrickvonplaten) [inline src/transformers/models/big_bird/modeling_flax_big_bird.py:1448]: Thanks for adding this!

File: src/transformers/pipelines/base.py [hunk 5]
```diff
@@ -1035,12 +856,27 @@ def forward(self, model_inputs, **forward_params):
...
     def get_iterator(
-        self, inputs, num_workers: int, batch_size: int, preprocess_params, forward_params, postprocess_params
+        self,
+        inputs,
+        num_workers: int,
+        batch_size: int,
+        preprocess_params,
+        forward_params,
+        postprocess_params,
+        align_to: Optional[int] = None,
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/base.py:879]: Does this mean it's "bucketed" to 8 or padded to 8? Not a huge fan of `align_to` - would prefer `bucket_to_multiple_of`

Comment (patrickvonplaten) [reply src/transformers/pipelines/base.py:879]: also maybe make this private in a first iteration to clearly state that it's experimental?

File: src/transformers/pipelines/fill_mask.py [hunk 6]
```diff
@@ -122,9 +128,19 @@ def postprocess(self, model_outputs, top_k=5, target_ids=None):
                 probs = probs[..., target_ids]
 
             values, predictions = probs.topk(top_k)
+        elif self.framework == "flax":
+            tokens = np.copy(input_ids)
+            masked_index = np.where(input_ids == self.tokenizer.mask_token_id)[0]
+            logits = outputs[0, masked_index.item(), :]
+            probs = jax.nn.softmax(logits, axis=0)
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/fill_mask.py:135]: if possible I would actually try to keep everything in numpy here to avoid automatically allocating TPU or GPU for the post-processing. I think numpy as both faster and safer here actually (especially for sorting stuff like topk) cc @patil-suraj

File: src/transformers/pipelines/image_classification.py [hunk 7]
```diff
@@ -87,8 +98,14 @@ def _forward(self, model_inputs):
...
         if top_k > self.model.config.num_labels:
             top_k = self.model.config.num_labels
-        probs = model_outputs.logits.softmax(-1)[0]
-        scores, ids = probs.topk(top_k)
+        if self.framework == "pt":
+            probs = model_outputs.logits.softmax(-1)[0]
+            scores, ids = probs.topk(top_k)
+        elif self.framework == "flax":
+            probs = softmax(model_outputs.logits)[0]
+            scores, ids = jax.lax.top_k(probs, top_k)
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/image_classification.py:106]: would probably also rather use numpy here...

File: src/transformers/pipelines/token_classification.py [hunk 8]
```diff
@@ -226,11 +227,22 @@ def _forward(self, model_inputs):
...
+        if self.framework == "flax":
+            logits = model_outputs["logits"][0]
+            special_tokens_mask = model_outputs["special_tokens_mask"][0]
+        else:
+            logits = model_outputs["logits"][0].numpy()
+            special_tokens_mask = model_outputs["special_tokens_mask"][0].numpy()
+
+        print("RECEIVED", special_tokens_mask)
+        if str(special_tokens_mask) == "1":
+            import ipdb
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/token_classification.py:239]: debugging left over

File: src/transformers/pipelines/token_classification.py [hunk 9]
```diff
@@ -264,8 +276,13 @@ def gather_pre_entities(
             # Filter special_tokens, they should only occur
             # at the sentence boundaries since we're not encoding pairs of
             # sentences so we don't have to keep track of those.
-            if special_tokens_mask[idx]:
-                continue
+            try:
+                if special_tokens_mask[idx]:
+                    continue
+            except Exception:
+                import ipdb
```
Comment (patrickvonplaten) [inline src/transformers/pipelines/token_classification.py:283]: same (I debug in the exact same way btw haha)

Comment (patrickvonplaten) [reply src/transformers/pipelines/token_classification.py:283]: try ... except + ipdb is epic

**Comment by patrickvonplaten** [comment]
> @patrickvonplaten I think this should be ready for a first pass review.
> 
> Let's focus on the big picture and overall architecture for this. The main pain points I am thinking of:
> 
> * Should we force padding to get compilation cache hit on jax.
> * Should we enable JIT automatically or not? (If we want `pjit` or others maybe we should delegate that choice to users ?)

Good questions! 

IMO:

For a) -> yes I would definitely do this. The by far most important application will be to provide a simple and fast demo for large LM models, such as https://huggingface.co/Cedille/fr-boris?text=Mon+nom+est+Thomas+et+mon+principal (which currently doesn't work with PT on GPU). I woulcd actually force padding to a large multiple to something like 64 by default (cc @patil-suraj what do you think?)

For b) -> yeah good question. I think we should delegate it to the user meaning that there will be both a `jit=True/False` and a `parallel=True/False` input that the user can specify. IMO both should default to False, but for the inference widget we should default jit to True I think.

File: src/transformers/feature_extraction_utils.py (hunk 1 above)
Comment (Narsil) [reply src/transformers/feature_extraction_utils.py:124]: We don't **need** that, but not having it forces in ALL pipelines to have

`return_framework="jax" if self.framework == "flax" else self.framework`
which is very tedious. 
As a user, I found it very confusing that all models are named `FlaxXXX` and this expected `jax`. 
Ofc I figured out how to solve, but seems unlikely (to me) that I would use `jax` tensors with `pythorch` models or anything like that (or use other tensor types within flax, even if numpy works)

Comment (Narsil) [reply src/transformers/feature_extraction_utils.py:124]: I understand what you're saying, but this is basically a way to ask for aliasing `flax` where possible (and where surprising)

File: src/transformers/generation_flax_utils.py (hunk 3 above)
Comment (Narsil) [reply src/transformers/generation_flax_utils.py]: Fair enough, I tend to prefer simpler lines usually but in that particular case I think you are right that inlined if is more readable.

File: src/transformers/models/big_bird/modeling_flax_big_bird.py (hunk 4 above)
Comment (Narsil) [reply src/transformers/models/big_bird/modeling_flax_big_bird.py:1448]: The main issue still exists when `padding` is not a multiple of `6 * block_size` which is suprising.

I didn't include in this PR since it's BigBird specific and it doesn't seem right to include this hardcoded `6 *` (and it fails other unrelated tests from `tests/test_flax_modeling_bigbird.py` since the seq_length used are NOT mutliples of `6 *`.

File: src/transformers/pipelines/token_classification.py (hunk 9 above)
Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:283]: Ahaha so useful in those very hard to detect error cases ;)

File: src/transformers/pipelines/token_classification.py (hunk 8 above)
Comment (Narsil) [reply src/transformers/pipelines/token_classification.py:239]: removed

File: src/transformers/pipelines/fill_mask.py (hunk 6 above)
Comment (Narsil) [reply src/transformers/pipelines/fill_mask.py:135]: Waiting for @patil-suraj 's confirmation on this.

In PT the difference should be pretty negligible in most cases, but I think doing the softmax on GPU *should* (afaiu) be faster in pipe mode since, you can keep the workload on the GPU, meaning less CPU work, meaning less downtime for the GPU (since `postprocess` happens on the main thread).

This would need some proved numbers IMO (for the pure optimization aspect)

File: src/transformers/pipelines/base.py [hunk 10]
```diff
@@ -62,80 +70,6 @@
 logger = logging.get_logger(__name__)
 
 
-def no_collate_fn(items):
```
[the remark this answers is not in the document]
Comment (Narsil) [reply src/transformers/pipelines/base.py:65]: There is a new `src/transformers/pipelines/utils.py` file added that contain those helpers.

`utils.py` would contain stuff not dependant on any particular framework (or their faillbacks) and `pt_utils.py` `flax_utils.py` and `tf_utils.py` could contain framework-dependant function.

File: src/transformers/pipelines/base.py (hunk 5 above)
Comment (Narsil) [reply src/transformers/pipelines/base.py:879]: It's not bucketing, it is forcing final `seq_length` to a specific mutliple of `align_to`.

Is `pad_to_multiple_of` more understandable to you ?

**Comment by Narsil** [comment]
> > @patrickvonplaten I think this should be ready for a first pass review.
> > Let's focus on the big picture and overall architecture for this. The main pain points I am thinking of:
> > 
> > * Should we force padding to get compilation cache hit on jax.
> > * Should we enable JIT automatically or not? (If we want `pjit` or others maybe we should delegate that choice to users ?)
> 
> Good questions!
> 
> IMO:
> 
> For a) -> yes I would definitely do this. The by far most important application will be to provide a simple and fast demo for large LM models, such as https://huggingface.co/Cedille/fr-boris?text=Mon+nom+est+Thomas+et+mon+principal (which currently doesn't work with PT on GPU). I woulcd actually force padding to a large multiple to something like 64 by default (cc @patil-suraj what do you think?)

Makes sense to me this was in the first iteration, and I removed it since during compilation during tests is still painfully slow because no cache hits there (we're mostly loading different models and sending a handful of inference at them).

> 
> For b) -> yeah good question. I think we should delegate it to the user meaning that there will be both a `jit=True/False` and a `parallel=True/False` input that the user can specify. IMO both should default to False, but for the inference widget we should default jit to True I think (if JAX is used for large models)

Ok, since this is specific to `jax` and I don't intend to add similar `torchscript` stuff to `pt` for instance, I am guessing I will aim for something that would look like

```python
pipe = pipeline(....)
pipe._enable_compilation(type="jit", )
# or
pipe._enable_compilation(type="pjit", **extra_mapping)
# Or add the jitting automatically if it wasn't called by the user with align to = 64
```

**Comment by patrickvonplaten** [comment]
@Narsil - that looks good to me. Let's see what @patil-suraj thinks about the API

**Comment by Narsil** [comment]
0Still valuable

**Comment by patil-suraj** [comment]
Super sorry about being so late here, will take a look this week :)

**Comment by patrickvonplaten** [comment]
Also think this would be a nice addition still - gently ping for @patil-suraj here.
